Abstract base Driver Class, follow up from #350

PhilipVinc opened the pull request:
@femtobit , @gcarleo 
This is a branched version of #350, (rebased on master, BTW). Since it changes quite a few things, I did not want to push to your branch.

Major changes compared to @femtobit PR are.
 - I created a Base Abstract Class `AbstractVariationalMonteCarlo` (name to be bikeshedded) which handles most IO, observables, run, iter.
    - Subclasses should implement the methods
        * `advance(self, n_steps)`
        * `estimate_stats(self, operator)`
        * `info(self, depth)`

All subclasses should then happily inherit all fancy functionality like Json output, progress bars, handling of observables...

 - I moved QSR, SteadyState and VMC to be sublcasses of the base class. We could in principle move more common stuff to the base class (sampler, machines...) but I'm not sure it's worth it.

 - If there is a quantity that is iteratively minimised, as in the case of VMC/Steadystate, then simply keeping `self._stats` up to date (As we currently do) will update the progress bar display and the JSON output. When this is null (as in QSR) this data is not printed/used.

Minor changes:
 - I moved all JSON logic to another file
 - I moved all `create_optimizer_fun` to another file, since they were now tri-plicated 
 - I implemented a very trivial `pytree_map` in pure python that handles lists and dicts. It can surely be improved.

I briefly tested this with `netket.variational.Vmc` and `netket.variational.steadystate` and it seems to work. If you like this setup, I could finish up the tests...
Moreover, I have not yet modified `variational.py` and `netket.Vmc` (as I think it will be deprecated). Making it a different constructor to `netket.variational.Vmc` seems the easier way to keep this deprecation around while reducing duplicated code.

PhilipVinc commented:
One of the errors related to tqdm I don't understand, but can delve into it...
One of the errors is because `netket.Vmc.iter(n_iters)` returns steps starting from 1 to `n_iters`, instead `netket.variational.Vmc.iter` used to start from 0. This can be easily fixed (though, are we aware that the two start iterating from different values?)

gcarleo commented:
The iterator should definitely start from 0 not 1, this is not Julia :) ahha

PhilipVinc commented:
But then whoever implemented ‘_Vmc.py’ was thinking about Julia because it starts from 1 (I just copy pasted it’s code)

PhilipVinc commented:
Can fix this quickly though.

femtobit commented on .envrc:
```diff
@@ -0,0 +1,13 @@
+layout_virtualenv() {
```
Committing your `.envrc` was accidental, right?

PhilipVinc replied:
yep, sorry.

femtobit commented:
> But then whoever implemented ‘_Vmc.py’ was thinking about Julia because it starts from 1 (I just copy pasted it’s code)

If I remember correctly, I chose to start at i=1 due to the fact that it feels kinda natural to use i=0 for the initial state and then i=1, ... for the simulation steps. This is especially true for time evolution, where ψ(i=0) and ψ(t=0) should be the same thing, IMHO. But staying consistent with the old drivers is also reasonable.

PhilipVinc commented:
I agree with @femtobit about the fact that it makes for sense starting at 1 is more consistent, but I'll let @gcarleo make the call.

An argument in favour of @gcarleo is that `iter` does not necessarily make sense for an adaptive time-stepping algorithm to solve the schroedinger's equation, because it might perform the solution for every time step (and it will probably return the time, and not the iteration number).

gcarleo commented:
These are certainly good arguments, but here I would invoke consistency with the rest of the language and start from 0. If you want we can add an option to start from an arbitrary index, like in ```range```

PhilipVinc commented:
This is not a 1 vs 0 indexing debate (I know this is Python).

It's about returning the step value before or after we have taken the step.
Consider the case where we iterate 2 steps at the time. What should be the steps returned? 
`[0,2,4,6,8...]` or `[2,4,6,8,10]` ? I advocate for the latter.  

```python
for iter in vmc.iter(n_steps=10, step=2):
   print(iter)
```

gcarleo commented:
It really depends on what we do when we compute the observables then: at the first iteration, do we compute the observables of the state before or after the gradient update? If it is after, then I also agree that starting from 1 is a better option.

PhilipVinc commented:
Right now what is done (and what this PR also does) is the following:
 - Compute Energy + Gradient
 - Update parameters
 - (eventually) compute observables

So observables and energy are always offset by a step.

It's a bit weird indeed that observables and energy/loss function are computed for different weights.

gcarleo commented:
Oups, then this is an issue, I guess that the energy and the other observables should be computed on the same wave function! Since the energy must be computed on the initial wf (to compute the gradient etc), I would suggest to also move the calculation of the observables to the state before the update (unless this is super tricky technically). In this way, the most natural thing to do would be to start iterations from 0. If we instead move the calculation of the energy to after the update, then we start from 1... @femtobit @PhilipVinc I don't have a strong preference on this, but I would just suggest to be consistent

femtobit commented:
Hm, yes, observables should be computed for the same weights as the energy. This is actually not that straightforward to do correctly.

The logic should probably be something like:

`for i = 0, ..., nsteps:`
1. Compute energy, gradient, and observables
2. Yield `i` with current data. For `i=0`, `parameters[0]` will be the unchanged initial parameters. (This is different from the behavior of the old driver, where `i=0` would correspond to the old energy but new parameters as @PhilipVinc has pointed out. But I think that makes a lot more sense.)
3. THEN, perform the update of the parameters `θ[i] -> θ[i+1]`.

gcarleo commented:
Yes, @femtobit that should be the logic, I agree. Is it a technical problem to perform the update after yielding?

PhilipVinc commented:
@femtobit But if we yield the parameters... it's a bit weird, no? we would lose the information that we already sampled the chain with which we can compute the observables.

@gcarleo Not really.Right now we compute gradient and perform the update step in the same function `vmc.advance`. 

I propose:
 - Define a new function `vmc.gradient()` that computes the energy/loss and the gradient (+sr), returning it. Move all the code in `vmc.advance` in there.

 - Change `vmc.advance` to match what @femtobit proposed. I would also add a callback argument that gets called after the gradient is computed but before the parameters are updated. We could yield the result of the callback...

femtobit commented:
I'd hope we can avoid callbacks. I think the current `advance` logic has to be split into a method for preparation (computing samples, energy, and gradient) and one for updating the state. For `iter`, the logic would then be (names just for illustration):
```
def iter(self, ...):
    for i in range(N):
        self.compute_samples_and_energy()
        yield i # Here, users can access self.energy and call self.estimate(obs)
        self.update_parameters()
```
`advance` could just be `compute_samples_and_energy` followed by `update_parameters`. The only problem is that `driver.energy` would then still be out of date.

PhilipVinc commented:
Ah right, `yield` in python is like a callback...
yes that's a good idea.

Yeah, `energy` would be out of sync, but I think this is fine if we document that properly?

femtobit commented:
> Yeah, energy would be out of sync, but I think this is fine if we document that properly?

I could live with that.

PhilipVinc commented:
Also, should we drop the possibility to have `step≠1` ?  
Otherwise... what should we yield? The only consistent choice would be the state before updating, so...

```python
    def iter(self, n_steps, step=1):
        for _ in range(0, n_steps, step):
            for i in range(0, step):
                dp = self.gradient()
                if i is 0 :
                    yield self.step_count
                    
                self.update_parameters(dp)
```

PhilipVinc commented:
@femtobit @gcarleo Please review, this is finished.

Notable changes now are the following:

  - Drivers:
    - The abstract base class `AbstractMCDriver` handles the creation of a default logger, the `tqdm` progress bar, and general interface.
     - The core logic of a driver should now be implemented in two methods called `_forward` and `_backward` performing the forward and backward passes. Alternatively you can implement just `_forward_and_bacward`, which is mostly equivalent to the old `vmc.advance(step=1)` method.
    - The other methods to be defined should be `reset`,  `info` and `estimate_stats`. The latter should return the stats object for a single operator. Handling of the pytrees is done in the base class.
    - The deprecated `netket.vmc` is now just a constructor with a different syntax of the same `netket.variational.vmc`, so that we have less code duplication.
    - The `vmc.run` method now accepts an optional keyword argument `obs` which should be a pytree (anything iterable, even lists) of observables. This is the preferred interface going forward

 - JsonLogger
   - I slightly changed the way that the JsonLogger works (@femtobit this might interest you). All the logic to compose the output is now in the driver, and the logger is just a sink. This should make it easier to add different loggers in the future. 
    - I also added a flush function and I call it at the end of `vmc.run()`. I had noticed that previously if you where saving every 30 steps but performing only 29 steps, this would not flush to file.
    - I added a lot of functionality to it (can append to files, overwrite, etc...). For now this functionality is not exported, but my idea is, in a following PR, to allow an user to create by himself the logger, and pass it as an argument to vmc drivers. (I'm thinking about appending to already existing logs, or composing multiple loggers like tensor board, etc...)

PhilipVinc commented:
@gcarleo can you restart the build that failed? 5/6 are fine, and that single one looks like a travis connection error to me. I'm convinced builds should be successful

femtobit commented:
@PhilipVinc Restarted it. And thanks for all the work on the driver, I'll give it a look.

femtobit commented on netket/logging.py:
```diff
@@ -0,0 +1,115 @@
+            if i is 1000:
+                raise IOError(
+                    "More than 1000 files exist with that prefix. Please change ouput_prefix manually."
+                )
```
Not sure whether this mode needs to be part of the logger code. I believe it's the responsibility of user code to provide the file name that should actually be used. We could provide this code here as a helper function, so that user who want his could call (my name might be too long - but descriptive :slightly_smiling_face:)
```python
prefix = netket.logging.increment_prefix_if_needed("myprefix")
vmc.run(..., prefix)
```

PhilipVinc replied:
So basically this is a choice between your syntax
```python
prefix = netket.logging.increment_prefix_if_needed("myprefix")
logger = JsonLogger(prefix)
```
and my syntax
```python
logger = JsonLogger(prefix, mode='newfile')
```
I am personally usually in favour of terse syntax. And in the future this can be moved to an `AbstractLogger` class.

It's not just a stylistic convenience: this same syntax is used by the TensorBoard loggers for Python (TensorBoardX) and Julia (TensorBoardLogger). By the way, In the latter it's the default behaviour, and I know people appreciate the fact that it's non-destructive by default.

femtobit replied:
Not sure something like `mode` should be part of an abstract logger interface, the meaning of it strongly depends on the output type (e.g., there could be a logger that writes to a remote database which has quite different semantics from a local file).

One issue I have with the current way: If I run a script that uses a constant prefix, say `test`, every time, I get the output files `test.log`, `test_1.log`, `test_2.log`, right? The first file would always break the pattern, which in my experience can be a bit annoying. User code however should know this and would be a good place to ensure a consistent naming for output.

PhilipVinc replied:
ok.
I'll remove this code for now, and leave the other three modes.

femtobit commented on netket/logging.py:
```diff
@@ -0,0 +1,115 @@
+    def __call__(self, step, item, machine):
+        item["Iteration"] = step
+
+        self._json_out["Output"].append(item)
```
In the longer term, it might be useful to modify this code so that it does not need to store the whole output (and thus have linearly growing memory usage over time) just to rewrite the file every time (that also increases the computational cost of the output step over time).

I think my old C++ JsonOutput class did only rewrite the necessary part at the end of the file, so that it was valid JSON all the time.

Doesn't have to be done in this PR, though.

PhilipVinc replied:
Yes, let's keep it for the future.

femtobit commented on netket/logging.py:
```diff
@@ -0,0 +1,115 @@
+
+    def _flush_log(self):
+        with open(self._prefix + ".log", "w") as outfile:
+            _json.dump(self._json_out, outfile)
```
Shouldn't this also only be done on MPI rank 0?

PhilipVinc replied:
My idea was that loggers are MPI-unaware, but the drivers knowingly only log/write/flush on rank 0. This simplifies implementing new loggers.

It's why I set `logger=None` in `abstractDriver.run`.

But if you think the opposite is more appropriate...

femtobit replied:
Ah, I see. But `_flush_params` does check the MPI rank right now.

I think generally I agree with having an MPI-unaware logger.

PhilipVinc replied:
Ops, copy-paste leftover. Will remove that check asap

femtobit commented on netket/abstract_driver.py:
```diff
@@ -0,0 +1,281 @@
+
+def obs_stat_to_dict(value):
+    st = value.asdict()
+    st["Mean"] = st["Mean"].real
```
I am unsure whether we should discard the imaginary part. In principle, defining a non-hermitian operator in NetKet should work in most places now otherwise, if I'm not mistaken.

PhilipVinc replied:
@gcarleo 

I agree with you that this is weird.
But, again, this is simply the code from the old drivers.
If we change this, this is a breaking change.

femtobit replied:
ok, true.

femtobit commented on netket/abstract_driver.py:
```diff
@@ -0,0 +1,281 @@
+import warnings
+
+
+def obs_stat_to_dict(value):
```
This should be internal, shouldn't it?
```suggestion
def _obs_stat_to_dict(value):
```

PhilipVinc commented on netket/_vmc.py:
```diff
@@ -186,130 +112,82 @@ def n_discard(self, n_discard):
 
-                grad = _mean(self._grads, axis=0)
+            # Center the local energy
+            eloc -= _mean(eloc)
```
I did not modify this code. I simply de-indented it. 
I can modify it, but thought that this was outside the scope of this PR...
@gcarleo

femtobit replied:
Fair enough, I'm fine with leaving it for now.

PhilipVinc commented on netket/abstract_driver.py:
```diff
@@ -0,0 +1,281 @@
+        pass
+
+    @abc.abstractmethod
+    def estimate_stats(self, observable):
```
Yes, good catch

PhilipVinc commented on netket/abstract_driver.py:
```diff
@@ -0,0 +1,281 @@
+        # flush at the end of the evolution so that final values are saved to
+        # file
+        if logger is not None:
+            logger.flush(self.machine)
```
TLDR: I sort of agree with you.
--
My intent for this PR was to do a PR with no breaking changes.
I have another PR sort-of ready with breaking changes (the argument order of `vmc.run`, for example). We can do this in the future PR.

The reason is that... JSON (in my opinion) is a shitty format. You cannot append to it. So at every step you are rewriting the full file, which eventually becomes very costly.  
Rather than flushing to file at every iteration (which you can do, just by passing the correct `save_every` flag), we could add some logic to JsonLogger to keep flushing as long as writing to file does not take more than XX% of total computational time.

femtobit replied:
Ok, then let's leave it in the current state for this PR.

I agree with JSON being sub-optimal for scientific data. Still, as an easy to read fallback format I think it beats ad-hoc CSV files or similar text output as it is at least structured.

For my own code I currently use a format where every line of the logfile is a JSON object (solves the appending problem and is still easy to process) for the high-level log info and [zarr](https://zarr.readthedocs.io/en/stable/) for larger array data (e.g., the machine parameters, correlators between all pairs of sites). That works quite well with `vmc.iter`.

PhilipVinc replied:
I was not aware of Zarr.
This might be nice.

PhilipVinc commented:
Ok, I think I addressed all your issues. 

For future reference, 
 - `_obs_state_to_dict` takes only the real part, and we should discuss if we want to log the full (Real + imaginary) parts, but that would be a breaking change of our output.
 - In `_vmc.py` it might be possible to remove the call to `mean` of the gradients.
 - The `Newfie` mode could be moved to an utility function
 - Flushing at predefined intervals of steps is bad. We should either find a way to append to json, change output format, or select an interval of time instead of steps.

For me this is good to go as soon as tests pass

gcarleo commented on netket/vmc_common.py:
```diff
@@ -0,0 +1,92 @@
+        return str(obj)
+
+
+def make_optimizer_fn(arg, ma):
```
Just for future reference, this should be completely removed in favor of an AbstractOptimizer class in python and possibly a wrapper for JAX/Tensorflow/Pytorch optimizers, as done for machines now

PhilipVinc commented:
After discussion with Giuseppe, I renamed `AbstractMCDriver` to `AbstractVariationalDriver` because drivers only store a machine and an optimiser (they don0't necessarily know anything about Monte Carlo sampling).

For GitHub automatic shenanigans, 
This PR
 - closes #345 (new `vmc.run` syntax with observables)
 - closes #348 (unify vmc and steadystate and qsr stuff)
 - closes #350 (completes that PR)

PhilipVinc commented:
I squashed the history a bit.
Merge ahead!
